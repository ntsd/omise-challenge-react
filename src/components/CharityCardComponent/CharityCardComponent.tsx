import React, {useState} from 'react';
import {Card, Button} from 'react-bootstrap';
import {Charity} from '../../types/Charity';
import {SelectAmountComponent} from '../SelectAmountComponent/SelectAmountComponent';

interface CharityCardProps {
  charity: Charity;
}

export const CharityCard: React.FC<CharityCardProps> = ({charity}) => {
  const [isSelectingAmount, setIsSelectingAmount] = useState(false);

  const CloseButton = () => (
    <Button
      className="close"
      aria-label="Close"
      style={{position: 'absolute', right: '20px', top: '20px'}}
      onClick={() => {
        setIsSelectingAmount(false);
      }}
    >
      <span aria-hidden="true">&times;</span>
    </Button>
  );

  const DonateButton = () => (
    <Button
      variant="outline-primary"
      onClick={() => {
        setIsSelectingAmount(true);
      }}
    >
      Donate
    </Button>
  );

  return (
    <Card style={{marginBottom: '50px'}}>
      {isSelectingAmount ? (
        <Card.Body style={{height: 'calc(35vh + 63px)', position: 'relative'}}>
          <CloseButton />
          <SelectAmountComponent charity={charity}></SelectAmountComponent>
        </Card.Body>
      ) : (
        <>
          <Card.Img
            variant="top"
            src={charity.image}
            style={{height: '35vh'}}
          />
          <Card.Footer
            className="d-flex justify-content-between"
            style={{backgroundColor: '#fff'}}
          >
            <div>{charity.name}</div>
            <DonateButton />
          </Card.Footer>
        </>
      )}
    </Card>
  );
};
