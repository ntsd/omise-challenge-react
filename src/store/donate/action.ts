export enum DonateActionType {
  UPDATE_TOTAL_DONATE = 'UPDATE_TOTAL_DONATE',
  UPDATE_MESSAGE = 'UPDATE_MESSAGE',
  REMOVE_MESSAGE = 'REMOVE_MESSAGE',
}
export type DonateAction = DonateActionType;

export interface DonateActionInterface {
  type: DonateAction;
  amount: number;
  message: string;
}
