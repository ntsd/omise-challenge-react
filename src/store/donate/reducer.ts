import {Reducer} from 'redux';
import {DonateActionInterface, DonateActionType} from './action';

export interface DonateState {
  donate: number;
  messages: string[];
}

const initialDonateState: DonateState = {
  donate: 0,
  messages: [],
};

export const DonateReducer: Reducer<DonateState, DonateActionInterface> = (
  state,
  action
) => {
  const _state = state || initialDonateState;

  switch (action.type) {
    case DonateActionType.UPDATE_TOTAL_DONATE:
      return Object.assign({}, _state, {
        donate: _state.donate + action.amount,
      });
    case DonateActionType.UPDATE_MESSAGE:
      return Object.assign({}, _state, {
        messages: [..._state.messages, action.message],
      });
    case DonateActionType.REMOVE_MESSAGE:
      return Object.assign({}, _state, {
        messages: _state.messages.slice(1),
      });
    default:
      return _state;
  }
};
