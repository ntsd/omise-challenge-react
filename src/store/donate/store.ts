import {createStore, Store} from 'redux';
import {DonateReducer, DonateState} from './reducer';

export const DonateStore: Store<DonateState> = createStore(DonateReducer);
