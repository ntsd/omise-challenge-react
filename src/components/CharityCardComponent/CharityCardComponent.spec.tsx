import React from 'react';
import renderer from 'react-test-renderer';
import {CharityCardComponent} from './CharityCardComponent';
import {DonateStore} from '../../store/donate/store';
import {Provider} from 'react-redux';

it('renders correctly', () => {
  const TestElement = () => {
    return (
      <Provider store={DonateStore}>
        <CharityCardComponent
          charity={{
            id: 1,
            name: 'test',
            image: 'image',
            currency: 'USD',
          }}
        ></CharityCardComponent>
      </Provider>
    );
  };

  const tree = renderer.create(<TestElement />).toJSON();
  expect(tree).toMatchSnapshot();
});
