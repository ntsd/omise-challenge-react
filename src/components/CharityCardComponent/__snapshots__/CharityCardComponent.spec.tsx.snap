// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="card"
  style={
    Object {
      "marginBottom": "50px",
    }
  }
>
  <img
    className="card-img-top"
    src="image"
    style={
      Object {
        "height": "35vh",
      }
    }
  />
  <div
    className="d-flex justify-content-between card-footer"
    style={
      Object {
        "backgroundColor": "#fff",
      }
    }
  >
    <div>
      test
    </div>
    <button
      className="btn btn-outline-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Donate
    </button>
  </div>
</div>
`;
